### PR TITLE
Add `mypy` and required typing changes because of it

### PR DESCRIPTION
**Pull request by ciaransweet**
This PR adds `mypy` to the project, which itself introduces the need to:

* Fix general typing errors
* Introduce new types in `chaosreliably.types` which represent the Reliably ObjectiveResults
* Marks the project as typed for use in other projects if needed
* Updates README & Makefile to cover off addition of `mypy`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -151,54 +151,67 @@ once every minute or less).
...
+[dco]: https://github.com/probot/dco#how-it-works
 
 ### Develop
 
 If you wish to develop on this project, make sure to install the development
 dependencies. But first, [create a virtual environment][venv] and then install
 those dependencies.
 
-[venv]: http://chaostoolkit.org/reference/usage/install/#create-a-virtual-environment
+[venv]: http://docs.chaostoolkit.org/reference/usage/install/#create-a-virtual-environment
```
Comment (CharlieMoon37) [inline README.md]: Tried your proposed link and it didn't seem to work for me whereas the original one did

```suggestion
[venv]: http://chaostoolkit.org/reference/usage/install/#create-a-virtual-environment
```

Comment (ciaransweet) [reply README.md]: That's what I get for copy pasting from other docs. CTK ones are wrong then haha

**Review by CharlieMoon37 — APPROVED** [review]
LGTM

**Review by Lawouach — APPROVED** [review]
Great work!

**Pull request merged**